### ISSUE TITLE
Issue with lowercase letters in Canadian postcode

### ISSUE THREAD

**Issue by chloe-meister** [dgilperez/validates_zipcode#56]
Canadian postal codes containing lower case letters are rejected as invalid:
```
ValidatesZipcode::InvalidZipcodeError (invalid zipcode V0H 1z6 for country CA)
```

I think this is because the [regexp for Canada](https://github.com/dgilperez/validates_zipcode/blob/a9a5803e88fb6087cabdb4dcf140fdb20aa79cb4/lib/validates_zipcode/cldr_regex_collection.rb#L17) only checks capital letters. This means that the postal code will be found invalid and [raise the error](https://github.com/dgilperez/validates_zipcode/blob/a9a5803e88fb6087cabdb4dcf140fdb20aa79cb4/lib/validates_zipcode/zipcode.rb#L20) before reaching the formatter, which capitalises the postal code.

Am I right? If yes I'd be happy to update the regexp and submit a PR.

**Comment by dgilperez** [comment]
Yes, that is correct.

Wikipedia claims that letters should be upcased, so it is the regex https://en.wikipedia.org/wiki/Postal_codes_in_Canada

But I agree that current implementation is wrong

```
def format
  raise InvalidZipcodeError, "invalid zipcode #{@zipcode} for country #{@country_alpha2.to_s.upcase}" unless valid?
  Formatter.new(zipcode: @zipcode, country_alpha2: @country_alpha2).format
end
```

`format` should call the formatter _before_ calling `valid?`, otherwise is not very useful. 

On the other hand, we might want to add the option for the validator to format first, right now they are unrelated. Something like `validates :zipcode, zipcode: { country_code: :es, format: true }`, maybe? 

If you'd like to open a PR, I'd gladly review.

**Comment by chloe-meister** [comment]
Sure, that'd be my pleasure to contribute!

**Comment by chloe-meister** [comment]
The format transformations in the Formatter actually assume the postcode is valid. For example, the formatter will format this invalid German postcode into a valid one:
```
ValidatesZipcode.format('KA9 1tr', 'DE', {format: true})
 => "00091" 
```
In this light, it makes sense to check the regexp before formatting the postal code to capture actual invalid postcodes. So I don't think the implementation is wrong after all!

In my case, many Canadian users seem to like to input lowercase letters instead of uppercase, so it's the regexp being too restrictive. If you'd agree with this I could update the regexp, basically adding `/i` at the end like it's done for the UK postcodes.

**Comment by dgilperez** [comment]
Hi @chloe-meister, sorry for the delay, hectic week. 

I am not sure here, let me walk it through with an example: 

* User inputs `V2r-1c8` as a CA code. 
* That will not be valid according to current regex, `/\A[ABCEGHJKLMNPRSTVXY]\d[ABCEGHJ-NPRSTV-Z][ ]?\d[ABCEGHJ-NPRSTV-Z]\d\z/`
* The formatter should turn that into a valid code, `V2R 1C8`
* If we call `valid?` before using the transformation, that will not be the case. 

So my question is ... why the specs are green if that's the case? https://github.com/dgilperez/validates_zipcode/blob/master/spec/format_zipcode_spec.rb#L6

`ValidatesZipcode::Formatter.new(zipcode: 'V2r-1c8', country_alpha2: 'CA').format` is now returning the formatted code, while `ValidatesZipcode.format('V2r-1c8', 'CA')` will raise an invalid error.

That is very inconsistent to say the least. 

I think the whole point to adding the format phase was to deal with loosely formatted codes added by users ... so I think we'd need to make those 2 consistent. 

What do you think?

**Comment by chloe-meister** [comment]
Hi @dgilperez,

Thanks for the answer and no worries for the delay 🙂 

I understand your point and you're right to point out that `ValidatesZipcode::Formatter.new(zipcode: 'V2r-1c8', country_alpha2: 'CA').format` and `ValidatesZipcode.format('V2r-1c8', 'CA')` should return the same result.

This currently also fails for Germany. `ValidatesZipcode::Formatter.new(zipcode: 'NG52DX', country_alpha2: 'DE').format` returns `00052` while `ValidatesZipcode.format('NG52DX', 'DE')` correctly throws an error. Here the issue is that the formatted postcode 00052 is invalid. 

For example, the following will never raise any error for German postcodes:
```ruby
formatted_zip = Formatter.new(zipcode: @zipcode, country_alpha2: @country_alpha2).format
valid = !!(regexp =~ formatted_zip)

raise InvalidZipcodeError, "invalid zipcode #{formatted_zip} for country #{@country_alpha2.to_s.upcase}" unless valid
```
because any non digit character will be replaced by 0 in formatting the postcode.

However the DE formatting is the only one using `rjust` so it's likely to be the only one affected by this issue.

What was the rationale behind replacing non-digits values by 0 in German postcodes? Maybe we can find an alternative way of formatting German postcode to avoid the issue above?

**Comment by dgilperez** [comment]
Closed via #57
